tls/s2n_handshake_io.c

```diff
@@ -107,14 +108,46 @@ static message_type_t handshakes[6][16] = {
 };
 
 #define ACTIVE_MESSAGE( conn ) handshakes[ (conn)->handshake.handshake_type ][ (conn)->handshake.message_number ]
+#define PREVIOUS_MESSAGE( conn ) handshakes[ (conn)->handshake.handshake_type ][ (conn)->handshake.message_number - 1 ]
+
 #define ACTIVE_STATE( conn ) state_machine[ ACTIVE_MESSAGE( (conn) ) ]
+#define PREVIOUS_STATE( conn ) state_machine[ PREVIOUS_MESSAGE( (conn) ) ]
 
 /* Used in our test cases */
 message_type_t s2n_conn_get_current_message_type(struct s2n_connection *conn)
 {
     return ACTIVE_MESSAGE(conn);
 }
 
+static int s2n_advance_message(struct s2n_connection *conn)
+{
+    char this = 'S';
+    if (conn->mode == S2N_CLIENT) {
+        this = 'C';
+    }
+
+    /* Actually advance the message number */
+    conn->handshake.message_number++;
+
+    /* Are we changing I/O directions */
+    if (ACTIVE_STATE(conn).writer == PREVIOUS_STATE(conn).writer) {
+        return 0;
+    }
+
+    /* We're the new writer */
+    if (ACTIVE_STATE(conn).writer == this) {
+        /* Set TCP_CORK/NOPUSH */
+        GUARD(s2n_socket_cork(conn));
```

Comment:
Should we worry about the opposite case? If the user has corked their socket prior to s2n_negotiate, we're going to clear that option on their fd afterwards